src/tile.cpp

```diff
@@ -1405,30 +1386,28 @@ void Tile::postRemoveNotification(Thing* thing, const Cylinder* newParent, int32
 		onUpdateTile(spectators);
 	}
 
-	for (Creature* spectator : spectators) {
-		assert(dynamic_cast<Player*>(spectator) != nullptr);
-		static_cast<Player*>(spectator)->postRemoveNotification(thing, newParent, index, LINK_NEAR);
+	for (const auto& spectator : spectators) {
+		assert(spectator->getPlayer() != nullptr);
+		std::static_pointer_cast<Player>(spectator)->postRemoveNotification(thing, newParent, index, LINK_NEAR);
 	}
 
 	// calling movement scripts
-	Creature* creature = thing->getCreature();
+	const auto& creature = thing->getCreature();
 	if (creature) {
-		g_moveEvents->onCreatureMove(creature, this, MOVE_EVENT_STEP_OUT);
+		g_moveEvents->onCreatureMove(creature, getTile(), MOVE_EVENT_STEP_OUT);
 	} else {
-		Item* item = thing->getItem();
+		const auto& item = thing->getItem();
 		if (item) {
-			g_moveEvents->onItemMove(item, this, false);
+			g_moveEvents->onItemMove(item, getTile(), false);
 		}
 	}
 }
 
-void Tile::internalAddThing(Thing* thing) { internalAddThing(0, thing); }
-
-void Tile::internalAddThing(uint32_t, Thing* thing)
+void Tile::internalAddThing(uint32_t, const std::shared_ptr<Thing>& thing)
 {
-	thing->setParent(this);
+	thing->setParent(getTile());
 
-	Creature* creature = thing->getCreature();
+	const auto& creature = thing->getCreature();
```

Comment:
`Creature` and `Item`.

```suggestion
	if (const auto& creature = thing->getCreature()) {
		g_game.map.clearSpectatorCache();
		if (creature->getPlayer()) {
			g_game.map.clearPlayersSpectatorCache();
		}

		creature->setParent(getTile());
		CreatureVector* creatures = makeCreatures();
		creatures->insert(creatures->begin(), creature);
	} else if (const auto& item = thing->getItem()) {
```